File: python/250_CountUnivalueTrees.py
class Solution:
    def countUnivalSubtrees(self, root: TreeNode) -> int:
        self.count = 0
        
        def helper(node):
            if not node:
                return True
            elif not node.left and not node.right:
                self.count += 1
                return True
            
            left, right = helper(node.left), helper(node.right)
            if left and right:
                if not node.left or not node.right:
                    if (node.left and node.left.val == node.val) or (node.right and node.right.val == node.val):
                        self.count += 1
                        return True
                    return False
                else:
                    if node.left.val == node.right.val == node.val:
                        self.count += 1
                        return True
            return False
        
        helper(root)
        return self.count

File: swift/250_CountUnivalueTrees.swift
class Solution {
    func countUnivalSubtrees(_ root: TreeNode?) -> Int {
        if root == nil { return 0 }
        
        var count: Int = 0
        
        func checkSubtree(_ node: TreeNode?, _ parentVal: Int) -> Bool {
            if node == nil { return true }
            
            var left: Bool = checkSubtree(node?.left, node!.val)
            var right: Bool = checkSubtree(node?.right, node!.val)
            
            if !left || !right { return false }
            
            count += 1
            return node!.val == parentVal
        }
        
        checkSubtree(root, root!.val)
        return count
    }
}
